### PR TITLE
[DS-3018] Add in multi-stage models

### PR DESCRIPTION
**Pull request by nathancooperjones**
## Describe the big picture - why are we doing this?
Okay... lots of changes here. 

As a bit of backstory, this PR started as a continuation of Hanna's hackweek, but she wasn't able to fully clean it up for a PR before she left. I took the branch over and, over the course of a straight 12 hour coding session, was able to clean it up for this PR. So here it is - a Hanna-Nate code mashup! 

The main changes here are the additions of the multi-stage models. These come in three files: one for establishing the base template class, one applying the template to a hybrid model, and another applying the template to a cold start model. Hanna initially had the cold start model for both users and items, but since we have another issue open to add user metadata to hybrid models (https://github.com/ShopRunner/collie_recs/issues/21), I decided to simplify the code a bit and just do item buckets for now. 

To make the multi-stage pipeline work with the existing `CollieMinimalTrainer`, I had to adjust how optimizers and learning rate schedulers work, which isn't a massive change. 

I also wanted to include bias terms into the hybrid model architectures, since Hanna did a big discussion on how important and beneficial they were to recommendations. 

So, with all that being said, here are the files I would look at: 

1. ``multi_stage_pipeline.py`` 
2. ``hybrid_pretrained_matrix_factorization.py`` 
3. ``hybrid_matrix_factorization.py`` 
4. ``cold_start_matrix_factorization.py``
5. Tutorial 06
6. Everything else. 

## Any additional details to clarify code in the PR?
Let me know if you see any potential or obvious mistakes in here, I have code-writing fatigue and might be making mistakes I don't realize yet. 

## How is this tested?
With a lot of unit tests! 

## Pull Request Checklist
 - [X] Pull request includes a description of why we are doing this
 - [X] ``CHANGELOG`` has been updated
 - [X] Version in ``_version.py`` has been updated
 - [X] All tests in the ``tests`` folder pass with a local build
 - [X] ``README`` has been updated (if applicable)
 - [X] Documentation in ``docs`` has been updated (if applicable)
 - [X] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)
 - [X] Docker image can be built using ``docker build -t collie_recs .``

### CODE REVIEW

File: collie_recs/model/cold_start_matrix_factorization.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        super().__init__(optimizer_config_list=optimizer_config_list,
+                         num_item_buckets=num_item_buckets,
+                         **get_init_arguments())
+
+    __doc__ = merge_docstrings(MultiStagePipeline, __doc__, __init__)
+
+    def _copy_weights(self, old: nn.Embedding, new: nn.Embedding, buckets: torch.tensor) -> None:
+        new.weight.data.copy_(old.weight.data[buckets])
+
+    def set_stage(self, stage: str) -> None:
```
Comment (sugi-chan) [inline collie_recs/model/cold_start_matrix_factorization.py:215]: so bc of the way the cold start models arer built these are the only two stages... cool

File: collie_recs/model/hybrid_pretrained_matrix_factorization.py [hunk 2]
```diff
@@ -245,16 +261,18 @@ def forward(self, users: torch.tensor, items: torch.tensor) -> torch.tensor:
...
         return pred_scores.squeeze()
 
-    def _get_item_embeddings(self) -> np.array:
+    def _get_item_embeddings(self) -> torch.tensor:
         """Get item embeddings."""
         # TODO: update this to get the embeddings post-MLP
-        return self.embeddings[1](
-            torch.arange(self.hparams.num_items, device=self.device)
-        ).detach().cpu()
+        return self.embeddings[1].weight.data
```
Comment (sugi-chan) [inline collie_recs/model/hybrid_pretrained_matrix_factorization.py:275]: ok so since this is happening everywhere im guessing its intentional., but we dont need to be detaching/moving things around on cpu and so on?

File: collie_recs/model/base/multi_stage_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                         **get_init_arguments())
+
+        if load_model_path is None:
+            # set stage if we have not already loaded it in and set it there
+            self.hparams.stage = self.hparams.stage_list[0]
+            self.set_stage(self.hparams.stage)
+
+    __doc__ = merge_docstrings(BasePipeline, __doc__, __init__)
+
+    def _load_model_init_helper(self, *args, **kwargs) -> None:
```
Comment (sugi-chan) [inline collie_recs/model/base/multi_stage_pipeline.py:123]: ok so this works because at this pt we have a trained model and want to load it in as if it was the last stage?

File: collie_recs/model/neural_collaborative_filtering.py [hunk 4]
```diff
@@ -194,11 +193,11 @@ def forward(self, users: torch.tensor, items: torch.tensor) -> torch.tensor:
...
         items = torch.arange(self.hparams.num_items, device=self.device)
 
-        return np.concatenate((
-            self.item_embeddings_cf(items).detach().cpu(),
-            self.item_embeddings_mlp(items).detach().cpu()
-        ), axis=1)
+        return torch.cat((
+            self.item_embeddings_cf(items),
+            self.item_embeddings_mlp(items),
+        ), axis=1).detach()
```
[the remark this answers is not in the document]
Comment (sugi-chan) [reply collie_recs/model/neural_collaborative_filtering.py:203]: i guess i commented on this twice at various pts

**Review by sugi-chan — CHANGES_REQUESTED** [review]
mostly just saying things look cool.... but have some questions and such. I will take another look through it tomorrow/later today lol

**Comment by nathancooperjones** [comment]
I can't actually tag @htorrence as a reviewer on this PR, but in case driving is really boring and you'd rather look at code for free, here you go!!

File: collie_recs/model/base/multi_stage_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+                 optimizer_config_list: List[Dict[str, Union[float, List[str], str]]] = None,
+                 loss: Union[str, Callable] = 'hinge',
+                 metadata_for_loss: Optional[Dict[str, torch.tensor]] = None,
+                 metadata_for_loss_weights: Optional[Dict[str, float]] = None,
+                 load_model_path: Optional[str] = None,
+                 map_location: Optional[str] = None,
+                 **kwargs):
+        stage_list = None
+
+        if optimizer_config_list is not None:
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/model/base/multi_stage_pipeline.py]: Yeah, this is not clear right now. It should only be `None` if we are loading in a model - I'll fix that now.

File: collie_recs/model/base/trainer.py [hunk 6]
```diff
@@ -349,7 +352,13 @@ def _train_loop_single_epoch(self, model: torch.nn.Module, epoch: int) -> float:
             loss = model.calculate_loss(batch)
             loss.backward()
 
-            self.optimizer.step()
+            for optimizer_idx, optimizer in enumerate(self.optimizer.optimizers):
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/model/base/trainer.py]: wait - are there optimizers that don't step on batches?

File: collie_recs/model/neural_collaborative_filtering.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie_recs/model/neural_collaborative_filtering.py]: I think here it's okay if these are on the device - they get used in the `item_item_similarity` method and we move it to CPU there.

File: tests/fixtures/model_fixtures.py [hunk 7]
```diff
@@ -279,8 +300,8 @@ def models_trained_for_one_step(request,
                                                   item_dense_layers_dims=[15, 10],
                                                   embedding_dropout_p=0.05,
                                                   dense_dropout_p=0.1,
-                                                  lr=1e-1,
-                                                  bias_lr=1e-2,
+                                                  lr=1e-2,
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply tests/fixtures/model_fixtures.py:303]: They should have been swapped from the getgo - typo!

File: collie_recs/model/base/multi_stage_pipeline.py (hunk 3 above)
Comment (nathancooperjones) [reply collie_recs/model/base/multi_stage_pipeline.py]: yes, exactly!

File: collie_recs/model/base/trainer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sugi-chan) [reply collie_recs/model/base/trainer.py]: oh wait brain was thinking badly lolll nvm

**Comment by sugi-chan** [comment]
I feel like doing PR reviews while driving could be bad

**Comment by nathancooperjones** [comment]
> I feel like doing PR reviews while driving could be bad

I mean, who's to say!

**Review by sugi-chan — APPROVED** [review]
I think this looks reasonable? /my brain is kinda going to mush after looking at it for awhile...

File: collie/model/base/multi_stage_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    __doc__ = merge_docstrings(BasePipeline, __doc__, __init__)
+
+    def _load_model_init_helper(self, *args, **kwargs) -> None:
+        super()._load_model_init_helper(*args, **kwargs)
+
+        # set the stage to the last stage
+        self.hparams.stage = self.hparams.stage_list[-1]
+        print(f'Set ``self.hparams.stage`` to "{self.hparams.stage}"')
+
+    def advance_stage(self) -> None:
```
Comment (sugi-chan) [inline collie/model/base/multi_stage_pipeline.py:136]: ok so was wondering what would happen if you like advanced past the len of the stage list or w/e... but I feel like thats kinda on a user... also since you call this it means you could in theory run a stage for more epochs if u needed to which is good. 

One thought would be if there was any interest in being able to go back a stage, but in theory a user could make use of the `set_tage` method which seems k

**Review by sugi-chan — APPROVED** [review]
tutorials seem good, I added a question about advancing through stages since someone might break it cuz ppl are bad? haha

File: tutorials/06_multi_stage_models.ipynb [hunk 9]
```diff
@@ -0,0 +1,2347 @@
+{
```
Comment (sugi-chan) [inline tutorials/06_multi_stage_models.ipynb:1]: <p>this is pretty great (setting all the seeds for nicer reproducibility is sweet)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ShopRunner/collie_recs/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (sugi-chan) [inline tutorials/06_multi_stage_models.ipynb:1]: <p><s>so in the first stage of training collie doesnt say the stage name, which it does in the following ones. I think that is a good quality of life improvement if not overly annoying to do.</s></p><p> </p><p>Ah ok it does get put higher up when initialized. I guess that works well. cool</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ShopRunner/collie_recs/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (sugi-chan) [inline tutorials/06_multi_stage_models.ipynb:1]: <p>nice addition to show why cold start is valuable</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ShopRunner/collie_recs/pull/22/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

File: collie/model/cold_start_matrix_factorization.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+        Returns
+        -------
+        preds: tensor, 1-d
+            Predicted ratings or rankings
+
+        """
+        user_embeddings = self.user_embeddings(users)
+        user_biases = self.user_biases(users)
+
+        if self.hparams.stage == 'item_buckets':
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie/model/cold_start_matrix_factorization.py:303]: I _totally_ agree - I feel like Octopod is a real natural candidate for this sort of thing.

File: collie/model/hybrid_pretrained_matrix_factorization.py [hunk 11]
```diff
@@ -236,8 +240,8 @@ def forward(self, users: torch.tensor, items: torch.tensor) -> torch.tensor:
             Predicted ratings or rankings
 
         """
-        if self.device != self.item_metadata.device:
-            self.item_metadata.to(self.device)
+        if str(self.device) != str(self.item_metadata.device):
+            self._move_any_external_data_to_device()
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie/model/hybrid_pretrained_matrix_factorization.py:244]: Eh, it's a bit hacky but the cleanest way I could do it for both `CollieTrainer` and `CollieMinimalTrainer`. Let me know if you think this is bad.

File: collie/model/cold_start_matrix_factorization.py [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+
+    def _copy_weights(self, old: nn.Embedding, new: nn.Embedding, buckets: torch.tensor) -> None:
+        new.weight.data.copy_(old.weight.data[buckets])
+
+    def set_stage(self, stage: str) -> None:
+        """Set the stage for the model."""
+        current_stage = self.hparams.stage
+
+        if stage in self.hparams.stage_list:
+            if current_stage == 'item_buckets' and stage == 'no_buckets':
```
[the remark this answers is not in the document]
Comment (nathancooperjones) [reply collie/model/cold_start_matrix_factorization.py:224]: yes, exactly!

File: collie/model/base/multi_stage_pipeline.py (hunk 8 above)
Comment (nathancooperjones) [reply collie/model/base/multi_stage_pipeline.py:136]: I agree, it should be on the user to not advance past the last stage, or else error. 

I think there shouldn't ever have to be a situation you reverse a stage (at least for now), but if someone needs to, they can just use `set_stage` for sure ¯\\_(ツ)_/¯
